src/python/pants/backend/python/lint/pylint/plugin_target_type.py

```diff
@@ -0,0 +1,63 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from pants.backend.python.target_types import COMMON_PYTHON_FIELDS, PythonSources
+from pants.engine.target import Dependencies, Target
+from pants.util.ordered_set import FrozenOrderedSet
+
+
+class PylintPluginSources(PythonSources):
+    required = True
+    expected_file_extensions = (".py",)
+
+
+class PylintPluginDependencies(Dependencies):
+    """Addresses to other targets that this plugin depends on.
+
+    These targets must either be third-party Python dependencies
+    (https://pants.readme.io/docs/python-third-party-dependencies) or be located within this
+    target's same directory or a subdirectory.
+    """
+
+
+class PylintSourcePlugin(Target):
+    """A Pylint plugin loaded through source code.
+
+    To load a source plugin:
+
+        1. Write your plugin. See http://pylint.pycqa.org/en/latest/how_tos/plugins.html.
+        2. Define a `pylint_source_plugin` target with the plugin's Python files included in the
+            `sources` field.
+        3. Add the parent directory of your target to the `root_patterns` option in the `[source]`
+            scope. For example, if your plugin is at `build-support/pylint/custom_plugin.py`, add
+            'build-support/pylint'. This is necessary for Pants to know how to tell Pylint to
+            discover your plugin. See https://pants.readme.io/docs/source-roots.
```

Comment:
It might be better to avoid this step by suggesting that folks declare it however they declare the rest of their code?

Reply:
They must declare a source root for the parent directory specifically. This is because the `load-plugins` Pylint option expects a module name, rather than the path to the module. You cannot specify `load-plugins=build-support/plugins/custom_plugin.py` - you must do `load-plugins=custom_plugin.py`.

Reply:
> This is because the load-plugins Pylint option expects a module name

This is the case any time you import code from another target... you always import via module name. So I don't think that differentiates this usecase.

Reply:
No, normally you would import `pants.util.strutil`. With Pylint, you cannot have the `pants.util` prefix. It must be just `strutil`.

Reply:
Meaning it doesn't support "dotted" module names? Hm. But the 3rdparty dependencies of the plugin could still be dotted presumably, because they would be hidden behind imports...?

Reply:
The top-level plugin cannot be dotted when configured with Pylint. When you use a third party plugin like `pylint_unittest`, you load it with `--load-plugins=pylint_unittest`.

In your source plugin's actual code, you can still import things that are dotted, such as the test plugin importing `subdir.dep`. The restriction is on what you pass to the `load-plugins` option.

Reply:
> In your source plugin's actual code, you can still import things that are dotted, such as the test plugin importing subdir.dep. The restriction is on what you pass to the load-plugins option.

Ok... but doesn't that mean that the restriction on dependency naming/location is not necessary?

Reply:
> but doesn't that mean that the restriction on dependency naming/location is not necessary?

You would think, if this were any normally well behaved Python program. But no, as described in https://github.com/pantsbuild/pants/pull/9792#discussion_r426885560.